File: agentprimordia/internal/agent/react_convert.go
// react_convert.go — 类型转换辅助函数
// 包含 Message → LLM ChatMessage 转换、多模态内容构建、ToolCall 转换等
package agent

import (
	"encoding/json"

	"agentprimordia/internal/llm"
)

// Helper functions for type conversion

func convertToLLMMessages(history []Message) []llm.ChatMessage {
	msgs := make([]llm.ChatMessage, 0, len(history))
	for _, m := range history {
		msg := llm.ChatMessage{
			Role:    string(m.Role),
			Content: m.Content,
		}

		if m.HasMultimodal() {
			msg.Content = buildMultimodalContent(m.ContentParts)
		}

		if m.Role == RoleAssistant && len(m.ToolCalls) > 0 {
			msg.ToolCalls = make([]llm.FunctionCall, len(m.ToolCalls))
			for j, tc := range m.ToolCalls {
				msg.ToolCalls[j] = llm.FunctionCall{
					ID:        tc.ID,
					Name:      tc.Name,
					Arguments: tc.Args,
				}
			}
		}
		if m.Role == RoleTool && len(m.Metadata.Extra) > 0 {
			if id, ok := m.Metadata.Extra["tool_call_id"]; ok {
				msg.ToolCallID = id
			}
			if isError, ok := m.Metadata.Extra["is_error"]; ok && isError == "true" {
				msg.IsToolError = true
			}
		}
		msgs = append(msgs, msg)
	}
	return msgs
}

// convertToOpenAIMessages 直接将 Message 切片转换为 OpenAI 兼容的 []map[string]any。
// 优化（Task 2.5）：消除历史中的双重转换（Message -> ChatMessage -> map[string]any），
// 在长对话（20+ 轮）中显著降低分配开销。
func convertToOpenAIMessages(history []Message) []map[string]any {
	msgs := make([]map[string]any, 0, len(history))
	for _, m := range history {
		content := m.Content
		if m.HasMultimodal() {
			content = buildMultimodalContent(m.ContentParts)
		}

		msg := map[string]any{
			"role":    string(m.Role),
			"content": content,
		}

		if m.Role == RoleAssistant && len(m.ToolCalls) > 0 {
			toolCalls := make([]map[string]any, len(m.ToolCalls))
			for j, tc := range m.ToolCalls {
				toolCalls[j] = map[string]any{
					"id":   tc.ID,
					"type": "function",
					"function": map[string]any{
						"name":      tc.Name,
						"arguments": tc.Args,
					},
				}
			}
			msg["tool_calls"] = toolCalls
		}

		if m.Role == RoleTool {
			if id, ok := m.Metadata.Extra["tool_call_id"]; ok {
				msg["tool_call_id"] = id
			}
		}

		msgs = append(msgs, msg)
	}
	return msgs
}

// convertToolDefsToLLMDefinitions 将 []map[string]any 形式的tool定义反解为 []llm.ToolDefinition。
// 优化（Task 2.5）：调用点直接持有 []llm.ToolDefinition 避免重复反解。
func convertToolDefsToLLMDefinitions(toolDefs []map[string]any) []llm.ToolDefinition {
	definitions := make([]llm.ToolDefinition, 0, len(toolDefs))
	for _, def := range toolDefs {
		fn, ok := def["function"].(map[string]any)
		if !ok {
			continue
		}
		typ, _ := def["type"].(string)
		name, _ := fn["name"].(string)
		desc, _ := fn["description"].(string)
		params, _ := fn["parameters"].(map[string]any)
		if name == "" {
			continue
		}
		definitions = append(definitions, llm.ToolDefinition{
			Type: typ,
			Function: llm.FunctionDefinition{
				Name:        name,
				Description: desc,
				Parameters:  params,
			},
		})
	}
	return definitions
}

// buildMultimodalContent 将 ContentParts 转换为 OpenAI 兼容的多模态 content JSON 字符串
func buildMultimodalContent(parts []ContentPart) string {
	type contentItem struct {
		Type     string `json:"type"`
		Text     string `json:"text,omitempty"`
		ImageURL *struct {
			URL    string `json:"url"`
			Detail string `json:"detail,omitempty"`
		} `json:"image_url,omitempty"`
	}

	items := make([]contentItem, len(parts))
	for i, p := range parts {
		switch p.Type {
		case "text":
			items[i] = contentItem{Type: "text", Text: p.Text}
		case "image_url":
			items[i] = contentItem{
				Type: "image_url",
				ImageURL: &struct {
					URL    string `json:"url"`
					Detail string `json:"detail,omitempty"`
				}{URL: p.URL, Detail: p.Detail},
			}
		case "image_b64":
			items[i] = contentItem{
				Type: "image_url",
				ImageURL: &struct {
					URL    string `json:"url"`
					Detail string `json:"detail,omitempty"`
				}{URL: "data:" + p.MIME + ";base64," + p.Data, Detail: p.Detail},
			}
		default:
			items[i] = contentItem{Type: "text", Text: p.Text}
		}
	}

	data, err := json.Marshal(items)
	if err != nil {
		type textItem struct {
			Type string `json:"type"`
			Text string `json:"text,omitempty"`
		}
		var fallback []textItem
		for _, p := range parts {
			if p.Text != "" {
				fallback = append(fallback, textItem{Type: "text", Text: p.Text})
			}
		}
		if fb, fbErr := json.Marshal(fallback); fbErr == nil {
			return string(fb)
		}
		return `[{"type":"text","text":""}]`
	}
	return string(data)
}

func convertToToolCalls(calls []llm.FunctionCall) []ToolCall {
	tcs := make([]ToolCall, len(calls))
	for i, c := range calls {
		tcs[i] = ToolCall{
			ID:   c.ID,
			Name: c.Name,
			Args: c.Arguments,
		}
	}
	return tcs
}

File: sdk/typescript/src/agent/react-convert.ts
// react-convert.ts — 消息类型转换工具
// 包含 Message → LLM ChatMessage、OpenAI 格式、工具定义等转换
// Mirrors Go internal/agent/react_convert.go

import type { Message, ToolCall, ToolDefinition } from '../types.js';

// ===== OpenAI 兼容类型 =====

export interface OpenAIMessage {
  role: 'system' | 'user' | 'assistant' | 'tool';
  content: string | OpenAIContentItem[];
  tool_calls?: OpenAIToolCall[];
  tool_call_id?: string;
  name?: string;
}

export interface OpenAIContentItem {
  type: string;
  text?: string;
  image_url?: {
    url: string;
    detail?: string;
  };
  input_audio?: {
    data: string;
    format: string;
  };
}

export interface OpenAIToolCall {
  id: string;
  type: 'function';
  function: {
    name: string;
    arguments: string;
  };
}

export interface OpenAIToolDefinition {
  type: 'function';
  function: {
    name: string;
    description: string;
    parameters: Record<string, unknown>;
  };
}

// ===== 多模态内容类型 =====

export interface ContentPart {
  type: 'text' | 'image_url' | 'image_b64' | 'audio';
  text?: string;
  url?: string;
  detail?: string;
  data?: string;
  mime?: string;
}

// ===== Message 扩展类型 =====

export interface ExtendedMessage extends Message {
  contentParts?: ContentPart[];
  metadata?: {
    extra?: Record<string, string>;
  };
}

// ===== 转换函数 =====

/** 将内部 Message 转换为 OpenAI 兼容格式 */
export function toOpenAIMessages(messages: Message[]): OpenAIMessage[] {
  return messages.map((m) => {
    const content = buildMultimodalContent(m);
    const msg: OpenAIMessage = {
      role: m.role,
      content,
    };

    // 工具调用
    if (m.role === 'assistant' && m.toolCalls && m.toolCalls.length > 0) {
      msg.tool_calls = m.toolCalls.map((tc) => ({
        id: tc.id,
        type: 'function' as const,
        function: {
          name: tc.name,
          arguments: tc.arguments,
        },
      }));
    }

    // 工具调用 ID
    if (m.role === 'tool' && m.toolCallId) {
      msg.tool_call_id = m.toolCallId;
    }

    // 名称
    if (m.name) {
      msg.name = m.name;
    }

    return msg;
  });
}

/** 构建多模态内容（兼容 OpenAI Vision API） */
export function buildMultimodalContent(m: Message): string | OpenAIContentItem[] {
  const ext = m as ExtendedMessage;
  if (!ext.contentParts || ext.contentParts.length === 0) {
    return m.content;
  }

  const items: OpenAIContentItem[] = ext.contentParts.map((p) => {
    switch (p.type) {
      case 'text':
        return { type: 'text', text: p.text ?? '' };
      case 'image_url':
        return {
          type: 'image_url',
          image_url: { url: p.url ?? '', detail: p.detail },
        };
      case 'image_b64':
        return {
          type: 'image_url',
          image_url: {
            url: `data:${p.mime ?? 'image/png'};base64,${p.data}`,
            detail: p.detail,
          },
        };
      case 'audio':
        return {
          type: 'input_audio',
          input_audio: {
            data: p.data ?? '',
            format: p.mime ?? 'wav',
          },
        };
      default:
        return { type: 'text', text: p.text ?? '' };
    }
  });

  return items;
}

/** 将 ToolDefinition 转换为 OpenAI 兼容格式 */
export function toOpenAIToolDefinitions(tools: ToolDefinition[]): OpenAIToolDefinition[] {
  return tools.map((t) => ({
    type: 'function' as const,
    function: {
      name: t.function.name,
      description: t.function.description,
      parameters: t.function.parameters,
    },
  }));
}

/** 将 OpenAI 工具调用转换回内部 ToolCall */
export function fromOpenAIToolCalls(calls: OpenAIToolCall[]): ToolCall[] {
  return calls.map((c) => ({
    id: c.id,
    name: c.function.name,
    arguments: c.function.arguments,
  }));
}

/** 将内部 ToolCall 转换为 OpenAI 工具调用 */
export function toOpenAIToolCalls(calls: ToolCall[]): OpenAIToolCall[] {
  return calls.map((c) => ({
    id: c.id,
    type: 'function' as const,
    function: {
      name: c.name,
      arguments: c.arguments,
    },
  }));
}

/** 从 OpenAI 消息中提取内部 Message */
export function fromOpenAIMessage(msg: OpenAIMessage): Message {
  const content = typeof msg.content === 'string' ? msg.content : JSON.stringify(msg.content);

  const message: Message = {
    role: msg.role as Message['role'],
    content,
  };

  if (msg.tool_calls && msg.tool_calls.length > 0) {
    message.toolCalls = fromOpenAIToolCalls(msg.tool_calls);
  }

  if (msg.tool_call_id) {
    message.toolCallId = msg.tool_call_id;
  }

  if (msg.name) {
    message.name = msg.name;
  }

  return message;
}

/** 从 OpenAI 消息列表中提取内部 Message 列表 */
export function fromOpenAIMessages(messages: OpenAIMessage[]): Message[] {
  return messages.map(fromOpenAIMessage);
}

// ===== 内容提取辅助函数 =====

/** 提取消息的纯文本内容 */
export function extractTextContent(msg: Message): string {
  const ext = msg as ExtendedMessage;
  if (ext.contentParts && ext.contentParts.length > 0) {
    return ext.contentParts
      .filter((p) => p.type === 'text')
      .map((p) => p.text ?? '')
      .join('\n');
  }
  return msg.content;
}

/** 检查消息是否包含多模态内容 */
export function hasMultimodal(msg: Message): boolean {
  const ext = msg as ExtendedMessage;
  return !!(ext.contentParts && ext.contentParts.length > 0);
}

/** 构建消息历史摘要（用于调试和日志） */
export function summarizeHistory(messages: Message[], maxChars: number = 200): string {
  const parts: string[] = [];
  let total = 0;

  for (const m of messages) {
    const text = extractTextContent(m);
    const line = `[${m.role}] ${text.slice(0, 50)}`;
    if (total + line.length > maxChars) {
      parts.push(`... (${messages.length - parts.length} more messages)`);
      break;
    }
    parts.push(line);
    total += line.length;
  }

  return parts.join('\n');
}